### PR TITLE
Separate UI from parsing before Stax porting

### PR DESCRIPTION
**Pull request by fbeutin-ledger**
Hello,
I have finished ported the Solana application for our latest device Ledger Stax
My work is taking place on top of the swap modifications I made
https://github.com/LedgerHQ/app-solana/pull/44

I have separated the new feature in three independant PRs:

- Add more Ragger tests for automated non regression testing and Stax screen reproducability
https://github.com/LedgerHQ/app-solana/pull/61

- Extract the UI code from the logic handling code for the current graphical library BAGL
https://github.com/LedgerHQ/app-solana/pull/62

- And finally the Stax porting itself for all screens
https://github.com/LedgerHQ/app-solana/pull/59

### CODE REVIEW

File: src/io.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#include "os.h"
+#include "ux.h"
+
+#include "apdu.h"
+#include "menu.h"
+#include "io.h"
+
+// override point, but nothing more to do
+void io_seproxyhal_display(const bagl_element_t *element) {
+    io_seproxyhal_display_default((bagl_element_t *) element);
```
Comment (xchapron-ledger) [inline src/io.c:30]: The cast can be removed.

File: src/menu.h [hunk 2]
```diff
@@ -1,9 +1,6 @@
+#pragma once
+
 #include "globals.h"
 #include "glyphs.h"
 
-#ifndef _MENU_H_
-#define _MENU_H_
-
 void ui_idle(void);
```
Comment (agrojean-ledger) [inline src/menu.h:6]: A bit weird that the header declaring the ui function is not in the ui subdirectory, isn't it ?

File: src/handle_sign_offchain_message.h [hunk 3]
```diff
@@ -1,10 +1,9 @@
...
 #include "cx.h"
 #include "globals.h"
 
-#ifndef _SIGN_OFFCHAIN_MESSAGE_H_
-#define _SIGN_OFFCHAIN_MESSAGE_H_
-
 void handle_sign_offchain_message(volatile unsigned int *flags, volatile unsigned int *tx);
 
-#endif
+void start_sign_offchain_message_ui(bool is_ascii, size_t num_summary_steps);
```
Comment (agrojean-ledger) [inline src/handle_sign_offchain_message.h:9]: Same thing, I find it weird not to have a dedicated ui header :thinking:

File: src/io.c [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            }
+
+        default:
+            THROW(ApduReplySdkInvalidParameter);
+    }
+    return 0;
+}
+
+void sendResponse(uint8_t tx, bool approve, bool display_menu) {
+    G_io_apdu_buffer[tx++] = approve ? 0x90 : 0x69;
```
Comment (agrojean-ledger) [inline src/io.c:92]: No defines for these codes ?

File: src/utils.h [hunk 5]
```diff
@@ -53,55 +53,11 @@ int read_derivation_path(const uint8_t *data_buffer,
...
-         (font) | BAGL_FONT_ALIGNMENT_CENTER,                           \
-         horizontalScrollSpeed},                                        \
-            (text), 0, 0, 0, NULL, NULL, NULL                           \
-    }
+uint8_t set_result_sign_message(void);
 
-#endif
+#endif  //_UTILS_H_
 
+// Outdated ?
```
Comment (agrojean-ledger) [inline src/utils.h:60]: So, outdated or not ? ^^

**Comment by xchapron-ledger** [comment]
It appears (CI failure) snapshots have changed? Was it wanted? Can you fix the CI?

File: src/menu.h (hunk 2 above)
Comment (fbeutin-ledger) [reply src/menu.h:6]: Moved to UI

File: src/io.c (hunk 1 above)
Comment (fbeutin-ledger) [reply src/io.c:30]: Done

File: src/menu.h (hunk 2 above)
Comment (fbeutin-ledger) [reply src/menu.h:6]: src/ui/ui_api.h with all UI calls

File: src/handle_sign_offchain_message.h (hunk 3 above)
Comment (fbeutin-ledger) [reply src/handle_sign_offchain_message.h:9]: src/ui/ui_api.h with all UI calls

File: src/io.c (hunk 4 above)
Comment (fbeutin-ledger) [reply src/io.c:92]: I would need to add `write_u16_be` and the likes and rework the call to have a proper function behavior, I'll let the devs handle this as it's not related to UI

File: src/utils.h (hunk 5 above)
Comment (fbeutin-ledger) [reply src/utils.h:60]: No idea, I think that it is but I'm not sure that the Solana devs are not using it

File: src/utils.h [hunk 6]
```diff
@@ -51,54 +53,11 @@ int read_derivation_path(const uint8_t *data_buffer,
...
-         0x000000,                                                      \
-         (font) | BAGL_FONT_ALIGNMENT_CENTER,                           \
-         horizontalScrollSpeed},                                        \
-            (text), 0, 0, 0, NULL, NULL, NULL                           \
-    }
-
-#endif
+#endif //_UTILS_H_
 
+// Outdated ?
```
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/utils.h:60]: No idea, I think that it's outdated but I'm not sure that the Solana devs are not using it

**Comment by fbeutin-ledger** [comment]
>  It appears (CI failure) snapshots have changed? Was it wanted? Can you fix the CI?

Should be fixed

**Review by agrojean-ledger — DISMISSED** [review]

**Pull request merged**